docs/architecture/adr-007.md

```diff
@@ -10,88 +13,101 @@ The current way to define those events is not flexible: events are hardcoded, pa
+- Add new argument `bytes extraData` to accept extra arguments in the future.
 
-```
+#### Signature
 
-- **functionName** corresponds to the equivalent hardcoded eventName such as `CronosSendToIBC` for the event `__CronosSendToIbc` or `CronosSendToChain` for the event `__CronosSendToChain`
-- **payer** is the address which make the payment, usually equal to `msg.sender`
-- **amount** corresponds to the `amount` that the payer has paid (either by burning or sending asset to the SC)
-- **args** corresponds to the argument in order of the event removing the `sender` and `amount` from the parameters encoded using the abi event pack
+```solidity
+// Current
+event __CronosSendToIbc(address sender, string recipient, uint256 amount);
 
+// Proposal
+event __CronosSendToIbcV2(address sender, string recipient, uint256 amount, string channelId, bytes extraData);
+```
 
+#### Description
 
-An example of smart contract implementing those function
+- Upgrade the event version for the argument changes
+  - Current event `__CronosSendToIbc` is kept to avoid introducing confusion and avoid extra chain logic to check the emitted event format at runtime.
+- Channel ID argument is added to support sending to specified IBC channel
+  - For token originated from Cronos, this channel ID could be any available IBC channel;
+  - For token originated from IBC channels, the initial implementation can add a restriction to enforce the same channel as the source channel before it is ready.
 
-```
-pragma solidity ^0.6.8;
+### __CronosSendToChain Event
 
-import "ds-token/token.sol";
+#### Changes
 
-Contract Demo is DSToken {
+- Change `chain_id` to `string` type to accept arbitrary chain ID;
+- Add new argument `extraData bytes` to accept extra arguments in the future.
 
- __CosmosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)
- 
- // unsafe_burn burn tokens without user's approval and authentication, used internally
- function unsafe_burn(address addr, uint amount) internal {
-     // Deduct user's balance without approval
-     require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
-     balanceOf[addr] = sub(balanceOf[addr], amount);
-     totalSupply = sub(totalSupply, amount);
-     emit Burn(addr, amount);
- }
- 
- // send an "amount" of the contract token to recipient through IBC
- // args should contains the encoded value recipient or extra value if needed in the future
- function send_to_ibc(uint amount, byte[] args) public {
-     unsafe_burn(msg.sender, amount);
-     emit __CosmosCallAfterPayment("CronosSendToIbc", msg.sender, amount, args);
- }
+#### Signature
 
+```solidity
+// Current
+event __CronosSendToChain(address sender, address recipient, uint256 amount, uint256 bridge_fee, uint256 chain_id);
+
+// Proposal
+event __CronosSendToChain(address sender, address recipient, uint256 amount, uint256 bridge_fee, string chain_id, extraData bytes);
```

Comment:
bridge_fee can be also encapsulated in extraData since its not clear what would be the format

Reply:
Sure, let me update it.